### ISSUE TITLE
Incorrect Custom Exception Handling (python)

### ISSUE THREAD

**Issue by nbanmp** [Z3Prover/z3#2078]
If a custom exception is raised during some parts of z3's execution, it will not always be handled gracefully. Perhaps there is something like an `except:` clause somewhere, catching every exception, rather than only the expected exceptions?

```python3
import signal
import time
from z3 import BitVec

class OutOfTimeError(Exception):
    pass

def sigalrm_handler(signum, frame):
    raise OutOfTimeError

def start_timeout(timeout):
    signal.signal(signal.SIGALRM, sigalrm_handler)
    signal.alarm(timeout)

def disable_timeout():
    signal.alarm(0)

# a is a random, long BitVec, so that `str(a)` needs to do a little more work.
a = BitVec('x', 256) + BitVec('x', 256) * BitVec('y', 256) + BitVec('z', 256) - BitVec('x', 256) / BitVec('x', 256)

try:
    start_timeout(1)
    time.sleep(0.998)
    str(a)
except OutOfTimeError:
    print("OutOfTimeError caught")
else:
    disable_timeout()
    print("Exception not caught")
```

Running this short script multiple times gives me three types of behaviors:
 1) Working correctly and the `OutOfTimeError` is caught.
 2) `Exception ignored in: <function AstRef.__del__ at 0x7f43f224bb70>` is printed, along with a traceback, and execution continues without catching the `OutOfTimeError`.
 3) `ctypes.ArgumentError: argument 1: <class '__main__.OutOfTimeError'>:` is raised, and execution halts immediately.

Note: this bug was found at https://github.com/ConsenSys/mythril-classic/pull/872

**Comment by NikolajBjorner** [comment]
Not sure how you would expect this to work.
From what I can see your test includes (at least) one race condition 
(because 1s is very close to .998 seconds) that is external to Z3: 
control reaches the else: branch. Before disable_timeout() takes effect it raises the signal.

**Comment by nbanmp** [comment]
Yes, there is a race condition to trigger this bug, but it is intentional. The goal of my script was to raise an exception during z3's execution, and this race condition was how I choose to do it.

A custom exception, in this case the OutOfTimeError, generally shouldn't be caught or result in a completely different exception by code that is not intended to handle it. This may be caused by using something like an `except:` instead of an `except SomeException`.

**Comment by nbanmp** [comment]
I don't know what the root cause of the issue is, it would likely require more debugging to figure out where the exception is being handled. If there are any `except:` clauses, it would generally be a good idea to replace them with `except <ExpectedException>:`, but that may or may not be the root cause of the errors my script reveals.

**Comment by apnewberry** [comment]
I'm not familiar with the Z3 internals, but perhaps I can mention some general Python folk-wisdom :-) 

Using `__del__` is generally error-prone and often discouraged. The [red box in the data model docs](https://docs.python.org/3/reference/datamodel.html#object.__del__) discusses some of the issues, including that "exceptions that occur during their execution are ignored".

Often when resource-cleanup is needed, a [context manager](https://www.python.org/dev/peps/pep-0343/) is the recommended option.
